nixos/lib/eval-config.nix

```diff
@@ -66,7 +66,7 @@ in rec {
   # These are the extra arguments passed to every module.  In
   # particular, Nixpkgs is passed through the "pkgs" argument.
   extraArgs = extraArgs_ // {
-    inherit baseModules extraModules modules;
+    inherit specialArgs baseModules extraModules modules;
```

Comment:
Doing it like this won't include the defaulted `modulesPath` in the `specialArgs` argument. Instead I suggest doing this in the `evalModules` definition instead:
```diff
diff --git a/lib/modules.nix b/lib/modules.nix
index 6b3faa447be..3ad1da4dd44 100644
--- a/lib/modules.nix
+++ b/lib/modules.nix
@@ -128,7 +128,7 @@ rec {
         let collected = collectModules
           (specialArgs.modulesPath or "")
           (modules ++ [ internalModule ])
-          ({ inherit lib options config; } // specialArgs);
+          ({ inherit lib options config specialArgs; } // specialArgs);
         in mergeModules prefix (reverseList collected);
 
       options = merged.matchedOptions;
```

This also has the advantage that no modules need to be evaluated to access `specialArgs` this way, which works the same for individual specialArgs values.

Reply:
My original reasoning was to actually not include `modulesPath` because eval-config sets a default for modulesPath. But I don't really have a reason for why thats a good thing. I switched to use your patch.

But with this is there any way to access `specialArgs` from an evaluated configuration? Right now you can do `nixosConfigurations.NixOS._module.args` to get all module args. And I would like to be able to access specialArgs that way too.

Reply:
Could we do something like this instead?
```
diff --git a/lib/modules.nix b/lib/modules.nix
index cc38e4cba87..97e3081c4b3 100644
--- a/lib/modules.nix
+++ b/lib/modules.nix
@@ -119,7 +119,7 @@ rec {
         };
 
         config = {
-          _module.args = args;
+          _module.args = args // { inherit specialArgs; };
         };
       };
```
I have no idea if that could break something.

Reply:
I switched to that method, let me know if its okay to do it like that

Reply:
I don't think this is a good idea, because it can cause `specialArgs` to be changed by modules, which then wouldn't match what they are used for. E.g. it would be possible to do `_module.args.specialArgs.foo = mkForce "something else"`, but the overridden value wouldn't apply to the actual module arguments.

Reply:
If you want to access `specialArgs` you could expose it from `eval-config.nix` instead

Reply:
Got it, yeah I see how that would be inconsistent. I switched back to your method. 

For my case I could just expose it inside a module: `{ specialArgs, ... }: { lib = { inherit specialArgs; }; }`. Then specialArgs can be accessed after. But thats a pretty rare use case, so I don't think it needs to be done by default. Its better than what I used to have to do: wrap eval-config to add specialArgs to `config.lib`.